File: hololive/hololive-rs/crates/alarm/service/src/dedup/fallback.rs
use std::{fmt::Display, time::Duration as StdDuration};

use alarm_core::constants::{LOCAL_FALLBACK_CLEANUP_MAX_KEYS, LOCAL_FALLBACK_DEDUP_TTL};
use moka::sync::Cache;

/// Valkey 장애 시 로컬 in-memory dedup 폴백.
///
/// - 키 존재 여부만 추적한다.
/// - TTL은 moka Cache의 `time_to_live`로 일괄 관리한다.
pub(crate) struct LocalDedupFallback {
    keys: Cache<String, ()>,
}

impl LocalDedupFallback {
    pub(crate) fn new() -> Self {
        Self {
            keys: Cache::builder()
                .max_capacity(LOCAL_FALLBACK_CLEANUP_MAX_KEYS as u64)
                .time_to_live(LOCAL_FALLBACK_DEDUP_TTL)
                .build(),
        }
    }

    /// Valkey SETNX 실패(outage) 시 로컬 폴백으로 claim을 시도한다.
    pub(crate) fn try_claim_on_outage<E: Display>(
        &self,
        key: &str,
        ttl: StdDuration,
        err: &E,
    ) -> bool {
        let fallback_ttl = normalize_fallback_ttl(ttl);
        let acquired = self.try_claim(key, fallback_ttl);
        tracing::warn!(
            key,
            fallback_acquired = acquired,
            error = %err,
            "SETNX claim 실패, 로컬 폴백 사용"
        );
        acquired
    }

    /// 로컬 dedup 클레임 시도.
    ///
    /// 현재 구현은 moka의 고정 TTL을 사용하므로 per-key TTL은 무시한다.
    fn try_claim(&self, key: &str, _ttl: StdDuration) -> bool {
        if self.keys.contains_key(key) {
            return false;
        }

        self.keys.insert(key.to_owned(), ());
        true
    }

    /// 로컬 dedup claim 해제.
    pub(crate) fn release_claims(&self, keys: &[String]) {
        for key in keys {
            self.keys.invalidate(key);
        }
    }
}

fn normalize_fallback_ttl(ttl: StdDuration) -> StdDuration {
    if ttl.is_zero() || ttl > LOCAL_FALLBACK_DEDUP_TTL {
        LOCAL_FALLBACK_DEDUP_TTL
    } else {
        ttl
    }
}

File: hololive/hololive-shared/pkg/service/alarm/dedup/fallback.go
// Copyright (c) 2025 Kapu
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package dedup

import (
	"log/slog"
	"sync"
	"sync/atomic"
	"time"

	"github.com/kapu/hololive-shared/pkg/constants"
)

// fallbackEntry: 로컬 dedup 엔트리
type fallbackEntry struct {
	expiresAtUnixNano int64
}

func (e fallbackEntry) isExpired(now time.Time) bool {
	return now.UnixNano() >= e.expiresAtUnixNano
}

type LocalFallback struct {
	logger   *slog.Logger
	keys     sync.Map
	keyCount atomic.Int64
	now      func() time.Time
}

type fallbackClaimResult struct {
	done     bool
	acquired bool
}

func NewLocalFallback(logger ...*slog.Logger) *LocalFallback {
	log := slog.Default()
	if len(logger) > 0 && logger[0] != nil {
		log = logger[0]
	}

	return &LocalFallback{
		logger: log,
		now:    time.Now,
	}
}

func (f *LocalFallback) TryClaimOnOutage(key string, ttl time.Duration, err error) bool {
	normalizedTTL := normalizeFallbackTTL(ttl)
	acquired := f.tryClaim(key, normalizedTTL)

	errMsg := ""
	if err != nil {
		errMsg = err.Error()
	}

	f.logger.Warn("SETNX claim 실패, 로컬 폴백 사용",
		slog.String("key", key),
		slog.Bool("fallback_acquired", acquired),
		slog.String("error", errMsg),
	)
	return acquired
}

func (f *LocalFallback) ReleaseClaims(claimKeys []string) {
	for _, key := range claimKeys {
		if _, loaded := f.keys.LoadAndDelete(key); loaded {
			f.keyCount.Add(-1)
		}
	}
}

// tryClaim: 로컬 dedup claim 시도
func (f *LocalFallback) tryClaim(key string, ttl time.Duration) bool {
	now := f.now()
	expiresAt := now.Add(ttl).UnixNano()

	// 용량 초과 시 만료 항목 정리
	if f.keyCount.Load() >= int64(constants.LocalFallbackCleanupMaxKeys) {
		f.cleanupExpired(now)
	}

	newEntry := fallbackEntry{expiresAtUnixNano: expiresAt}

	for {
		result := f.tryClaimOnce(key, now, newEntry)
		if result.done {
			return result.acquired
		}
	}
}

func (f *LocalFallback) tryClaimOnce(key string, now time.Time, newEntry fallbackEntry) fallbackClaimResult {
	loadedEntry, exists := f.keys.Load(key)
	if !exists {
		return fallbackClaimResult{done: f.tryStoreNewClaim(key, newEntry), acquired: true}
	}

	entry, ok := loadedEntry.(fallbackEntry)
	if !ok {
		f.deleteCorruptEntry(key, loadedEntry)
		return fallbackClaimResult{}
	}

	if !entry.isExpired(now) {
		return fallbackClaimResult{done: true}
	}

	return fallbackClaimResult{
		done:     f.keys.CompareAndSwap(key, entry, newEntry),
		acquired: true,
	}
}

func (f *LocalFallback) tryStoreNewClaim(key string, newEntry fallbackEntry) bool {
	if _, loaded := f.keys.LoadOrStore(key, newEntry); loaded {
		return false
	}
	f.keyCount.Add(1)
	return true
}

func (f *LocalFallback) deleteCorruptEntry(key any, value any) {
	if f.keys.CompareAndDelete(key, value) {
		f.keyCount.Add(-1)
	}
}

// cleanupExpired: 만료된 항목 정리 (잠금 보유 상태에서 호출)
func (f *LocalFallback) cleanupExpired(now time.Time) {
	f.keys.Range(func(key, value any) bool {
		entry, ok := value.(fallbackEntry)
		if !ok {
			f.deleteCorruptEntry(key, value)
			return true
		}

		if entry.isExpired(now) {
			f.deleteExpiredEntry(key, entry)
		}
		return true
	})
}

func (f *LocalFallback) deleteExpiredEntry(key any, entry fallbackEntry) {
	if f.keys.CompareAndDelete(key, entry) {
		f.keyCount.Add(-1)
	}
}

// normalizeFallbackTTL: TTL 정규화 (0 또는 최대치 초과 시 기본값 적용)
func normalizeFallbackTTL(ttl time.Duration) time.Duration {
	if ttl == 0 || ttl > constants.LocalFallbackDedupTTL {
		return constants.LocalFallbackDedupTTL
	}
	return ttl
}
